src/main/java/io/vertx/core/impl/ContextBase.java

```diff
@@ -36,15 +37,19 @@ public abstract class ContextBase implements ContextInternal {
 
   private static final String DISABLE_TIMINGS_PROP_NAME = "vertx.disableContextTimings";
   static final boolean DISABLE_TIMINGS = Boolean.getBoolean(DISABLE_TIMINGS_PROP_NAME);
+  private static final AtomicReferenceFieldUpdater<ContextBase, ConcurrentMap> DATA_UPDATER =
+    AtomicReferenceFieldUpdater.newUpdater(ContextBase.class, ConcurrentMap.class, "data");
+  private static final AtomicReferenceFieldUpdater<ContextBase, ConcurrentMap> LOCAL_DATA_UPDATER =
+    AtomicReferenceFieldUpdater.newUpdater(ContextBase.class, ConcurrentMap.class, "localData");
 
   private final VertxInternal owner;
   private final JsonObject config;
   private final Deployment deployment;
   private final CloseFuture closeFuture;
   private final ClassLoader tccl;
   private final EventLoop eventLoop;
-  private ConcurrentMap<Object, Object> data;
-  private ConcurrentMap<Object, Object> localData;
+  private volatile ConcurrentMap<Object, Object> data;
+  private volatile ConcurrentMap<Object, Object> localData;
```

Comment:
I am wondering why you have made these fields volatile here while not doing the same for `DuplicatedContext`

Reply:
Because I've missed it for sure, thanks!